### ISSUE TITLE
LLD from LLVM 12.0.0 can not link binaries on virtual disk

### ISSUE THREAD

**Issue by luisdallos** [brechtsanders/winlibs_mingw#66]
This issue is similar to the one reported by **wirx6** at msys2/MINGW-packages#8664

After updating from [`winlibs-x86_64-posix-seh-gcc-10.2.0-llvm-11.0.0-mingw-w64-8.0.0-r8.7z`](/brechtsanders/winlibs_mingw/releases/download/10.2.0-11.0.0-8.0.0-r8/winlibs-x86_64-posix-seh-gcc-10.2.0-llvm-11.0.0-mingw-w64-8.0.0-r8.7z) (tag: [`10.2.0-11.0.0-8.0.0-r8`](/brechtsanders/winlibs_mingw/releases/tag/10.2.0-11.0.0-8.0.0-r8)) to [`winlibs-x86_64-posix-seh-gcc-10.3.0-llvm-12.0.0-mingw-w64-9.0.0-r2.7z`](/brechtsanders/winlibs_mingw/releases/download/10.3.0-12.0.0-9.0.0-r2/winlibs-x86_64-posix-seh-gcc-10.3.0-llvm-12.0.0-mingw-w64-9.0.0-r2.7z) (tag: [`10.3.0-12.0.0-9.0.0-r2`](/brechtsanders/winlibs_mingw/releases/tag/10.3.0-12.0.0-9.0.0-r2)), it is impossible to link binaries with lld if output file is located on virtual disk. Problem is not occurring after replacing both `ld.lld.exe` and `lld-link.exe` binaries with the ones from LLVM 11.0.0. Same situation with the 32-bit (i.e., `i686-posix-dwarf-gcc`) toolchain.

Steps to Reproduce:

- Create a virtual disk with OSFMount or any other virtual disk software
- `cd` to virtual disk filesystem
- `echo "int main(){}" > test.c`
- `clang -fuse-ld=lld -o test.exe test.c`

Logs: (lld 12.0.0 + bfd + lld 11.0.0):

- LLD 12.0.0:
```
$ ld.lld --version
LLD 12.0.0 (compatible with GNU linkers)

$ clang -v -fuse-ld=lld -o test-lld.exe test.c
(built by Brecht Sanders) clang version 12.0.0
Target: x86_64-w64-windows-gnu
Thread model: posix
InstalledDir: J:\mingw64\bin
 "J:\\mingw64\\bin\\clang.exe" -cc1 -triple x86_64-w64-windows-gnu -emit-obj -mrelax-all --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name test.c -mrelocation-model pic -pic-level 2 -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -mms-bitfields -munwind-tables -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -v -resource-dir "J:\\mingw64\\lib\\clang\\12.0.0" -internal-isystem "J:\\mingw64\\lib\\clang\\12.0.0\\include" -internal-isystem "J:\\mingw64\\x86_64-w64-mingw32/sys-root/mingw/include" -internal-isystem "J:\\mingw64\\x86_64-w64-mingw32\\include" -internal-isystem "J:\\mingw64\\include" -fdebug-compilation-dir "G:\\" -ferror-limit 19 -fno-use-cxa-atexit -fgnuc-version=4.2.1 -exception-model=seh -faddrsig -o "G:\\tmp\\test-ec23cb.o" -x c test.c
clang -cc1 version 12.0.0 based upon LLVM 12.0.0 default target x86_64-w64-mingw32
ignoring nonexistent directory "J:\mingw64\x86_64-w64-mingw32/sys-root/mingw/include"
#include "..." search starts here:
#include <...> search starts here:
 J:\mingw64\lib\clang\12.0.0\include
 J:\mingw64\x86_64-w64-mingw32\include
 J:\mingw64\include
End of search list.
 "J:\\mingw64\\bin\\ld.lld" -m i386pep -Bdynamic -o test-lld.exe "J:\\mingw64\\x86_64-w64-mingw32\\lib\\crt2.o" "J:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0\\crtbegin.o" "-LJ:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0" "-LJ:\\mingw64\\x86_64-w64-mingw32\\lib" "-LJ:\\mingw64\\lib" "-LJ:\\mingw64\\x86_64-w64-mingw32/sys-root/mingw/lib" "-LJ:\\mingw64\\lib\\clang\\12.0.0\\lib\\windows" "G:\\tmp\\test-ec23cb.o" -lmingw32 -lgcc -lgcc_eh -lmoldname -lmingwex -lmsvcrt -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc -lgcc_eh -lmoldname -lmingwex -lmsvcrt -lkernel32 "J:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0\\crtend.o"
lld-link: error: cannot open output file test-lld.exe: Function not implemented
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

- BFD 2.36.1:
```
$ ld.bfd --version
GNU ld (Binutils for MinGW-W64 x86_64, built by Brecht Sanders) 2.36.1

$ clang -v -fuse-ld=bfd -o test-bfd.exe test.c
(built by Brecht Sanders) clang version 12.0.0
Target: x86_64-w64-windows-gnu
Thread model: posix
InstalledDir: J:\mingw64\bin
 "J:\\mingw64\\bin\\clang.exe" -cc1 -triple x86_64-w64-windows-gnu -emit-obj -mrelax-all --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name test.c -mrelocation-model pic -pic-level 2 -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -mms-bitfields -munwind-tables -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -v -resource-dir "J:\\mingw64\\lib\\clang\\12.0.0" -internal-isystem "J:\\mingw64\\lib\\clang\\12.0.0\\include" -internal-isystem "J:\\mingw64\\x86_64-w64-mingw32/sys-root/mingw/include" -internal-isystem "J:\\mingw64\\x86_64-w64-mingw32\\include" -internal-isystem "J:\\mingw64\\include" -fdebug-compilation-dir "G:\\" -ferror-limit 19 -fno-use-cxa-atexit -fgnuc-version=4.2.1 -exception-model=seh -faddrsig -o "G:\\tmp\\test-fbd123.o" -x c test.c
clang -cc1 version 12.0.0 based upon LLVM 12.0.0 default target x86_64-w64-mingw32
ignoring nonexistent directory "J:\mingw64\x86_64-w64-mingw32/sys-root/mingw/include"
#include "..." search starts here:
#include <...> search starts here:
 J:\mingw64\lib\clang\12.0.0\include
 J:\mingw64\x86_64-w64-mingw32\include
 J:\mingw64\include
End of search list.
 "J:\\mingw64\\bin\\x86_64-w64-mingw32-ld.bfd" -m i386pep -Bdynamic -o test-bfd.exe "J:\\mingw64\\x86_64-w64-mingw32\\lib\\crt2.o" "J:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0\\crtbegin.o" "-LJ:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0" "-LJ:\\mingw64\\x86_64-w64-mingw32\\lib" "-LJ:\\mingw64\\lib" "-LJ:\\mingw64\\x86_64-w64-mingw32/sys-root/mingw/lib" "-LJ:\\mingw64\\lib\\clang\\12.0.0\\lib\\windows" "G:\\tmp\\test-fbd123.o" -lmingw32 -lgcc -lgcc_eh -lmoldname -lmingwex -lmsvcrt -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc -lgcc_eh -lmoldname -lmingwex -lmsvcrt -lkernel32 "J:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0\\crtend.o"
```

- LLD 11.0.0:
```
$ ld.lld --version
LLD 11.0.0 (compatible with GNU linkers)

$ clang -v -fuse-ld=lld -o test-lld.exe test.c
(built by Brecht Sanders) clang version 12.0.0
Target: x86_64-w64-windows-gnu
Thread model: posix
InstalledDir: J:\mingw64\bin
 "J:\\mingw64\\bin\\clang.exe" -cc1 -triple x86_64-w64-windows-gnu -emit-obj -mrelax-all --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name test.c -mrelocation-model pic -pic-level 2 -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -mms-bitfields -munwind-tables -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -v -resource-dir "J:\\mingw64\\lib\\clang\\12.0.0" -internal-isystem "J:\\mingw64\\lib\\clang\\12.0.0\\include" -internal-isystem "J:\\mingw64\\x86_64-w64-mingw32/sys-root/mingw/include" -internal-isystem "J:\\mingw64\\x86_64-w64-mingw32\\include" -internal-isystem "J:\\mingw64\\include" -fdebug-compilation-dir "G:\\" -ferror-limit 19 -fno-use-cxa-atexit -fgnuc-version=4.2.1 -exception-model=seh -faddrsig -o "G:\\tmp\\test-909d99.o" -x c test.c
clang -cc1 version 12.0.0 based upon LLVM 12.0.0 default target x86_64-w64-mingw32
ignoring nonexistent directory "J:\mingw64\x86_64-w64-mingw32/sys-root/mingw/include"
#include "..." search starts here:
#include <...> search starts here:
 J:\mingw64\lib\clang\12.0.0\include
 J:\mingw64\x86_64-w64-mingw32\include
 J:\mingw64\include
End of search list.
 "J:\\mingw64\\bin\\ld.lld" -m i386pep -Bdynamic -o test-lld.exe "J:\\mingw64\\x86_64-w64-mingw32\\lib\\crt2.o" "J:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0\\crtbegin.o" "-LJ:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0" "-LJ:\\mingw64\\x86_64-w64-mingw32\\lib" "-LJ:\\mingw64\\lib" "-LJ:\\mingw64\\x86_64-w64-mingw32/sys-root/mingw/lib" "-LJ:\\mingw64\\lib\\clang\\12.0.0\\lib\\windows" "G:\\tmp\\test-909d99.o" -lmingw32 -lgcc -lgcc_eh -lmoldname -lmingwex -lmsvcrt -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc -lgcc_eh -lmoldname -lmingwex -lmsvcrt -lkernel32 "J:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\10.3.0\\crtend.o"
```

**Comment by brechtsanders** [comment]
The right place to report this would be to the LLVM development team, so they can fix the issue in their code.
Has it been reported there?

**Comment by brechtsanders** [comment]
I had similar issues because I'm using an ImDisk RAM-drive to speed up builds.
This issue has already been reported here: https://bugs.llvm.org/show_bug.cgi?id=52080

**Comment by luisdallos** [comment]
https://github.com/mstorsjo/llvm-mingw/issues/248

Closing this issue as it has been fixed for good in LLVM 14.0.0
